homeassistant/components/tplink/common.py

```diff
@@ -100,6 +98,37 @@ def process_devices() -> None:
             else:
                 _LOGGER.error("Unknown smart device type: %s", type(dev))
 
+    discovery_attempt = 0
+    devices = dict()
+
+    while discovery_attempt < MAX_DISCOVERY_RETRIES:
+        discovery_attempt += 1
+        _LOGGER.debug(
+            "Discovering tplink devices, attempt %s of %s",
+            discovery_attempt,
+            MAX_DISCOVERY_RETRIES,
+        )
+        discovered_devices = await async_get_discoverable_devices(hass)
+        _LOGGER.info(
+            "Discovered %s TP-Link of expected %s smart home device(s)",
+            len(discovered_devices),
+            target_discovery_qty,
+        )
+        for ip in discovered_devices:
+            if ip not in devices:
+                devices[ip] = discovered_devices[ip]
```

Comment:
This could be simplified by not checking if the device already exist - the backend lib will create instances based on discovery responses anyway, so this doesn't optimize the process.